src/test/java/lambda/part3/exercise/ReachIterableTest.java

```diff
@@ -46,32 +46,41 @@
             );
         }
 
-        default <R> ReachIterable<R> flatMap(final Function<T, List<R>> function) {
-            final List<R> results = new ArrayList<>();
-            while (this.forNext(element -> results.addAll(function.apply(element))));
-            return ReachIterable.from(results);
+        default <R> ReachIterable<R> flatMap(final Function<T, List<R>> mapper) {
+            final AtomicReference<ReachIterable<R>> inner = new AtomicReference<>(
```

Comment:
Atomics are both expensive and misleading. These classes supposed to be used in thread safe code.

You could create private class in test class or in anonymous class.